[WIP] Correctly checkout rev when updating mix.exs

whatyouhide opened the pull request:
This is an attempt to fix #4309. I'm a little new to the `Mix.Dep*` part of the codebase, so I'm not sure this is a good solution; also, this is of course a work-in-progress as it completely lacks tests, but I wanted to see if the direction is right before moving on. Let me know what I can do :)

josevalim commented on lib/mix/lib/mix/scm/git.ex:
```diff
@@ -52,11 +52,12 @@ defmodule Mix.SCM.Git do
         File.cd!(opts[:dest], fn ->
           rev_info = get_rev_info
```
Can we get the tag information here? I would like to avoid multiple calls into git.

whatyouhide replied:
@josevalim I implemented this (calling git only once) in aa5276f; I think the implementation looks a little bit more noisy, but I agree this may be worth it.

josevalim commented on lib/mix/lib/mix/scm/git.ex:
```diff
@@ -155,16 +156,25 @@ defmodule Mix.SCM.Git do
-    destructure [origin, rev],
-      :os.cmd('git --git-dir=.git config remote.origin.url && git --git-dir=.git rev-parse --verify --quiet HEAD')
+  defp get_rev_info(opts) do
+    cmd = 'git --git-dir=.git config remote.origin.url && git --git-dir=.git rev-parse --verify --quiet HEAD'
```
Could we make the second command simply:

    rev_in_opts = opts[:tag] || opts[:branch] || opts[:rev] || "HEAD"
    cmd = "git --git-dir=.git rev-parse --verify --quiet #{rev_in_opts}~0"

Or do we need the third command? I think we just need to get the proper rev once.

If we need a third command, notice that you need to pass `--git-dir=.git` to the command you added below. :)

josevalim replied:
Actually, I think it should be this:

    defp get_rev_info(rev) do
      cmd = 'git --git-dir=.git config remote.origin.url && git --git-dir=.git rev-parse --verify --quiet #{rev}~0'
      # same as today
    end

Then, in get_lock, we will call:

    get_rev_info("HEAD")

but when checking the lock:

    get_rev_info(opts[:tag] || opts[:branch] || opts[:rev] || "HEAD")

This way we don't need to add this line up above:

    old_rev_as_of_opts?(rev_info, lock_rev)

Because the following line will always do the proper check:

    lock_rev  != rev_info[:rev] 

Does this make sense?

josevalim replied:
Actually, it doesn't make sense. We always need to check HEAD because that's where we actually point to. There is something fishy about this still, let me see if I can put a finger on it.

josevalim replied:
Ok, my concern with the code above is that today it is completely fine to depend on an earlier commit in a branch. For example, imagine you have branch: "master", we don't need to depend on the latest commit in master but this change will always push to the latest commit in there. Therefore, I think the checks we have added are not good, because it is ok to not be in the latest rev in a branch. However, we need to make sure that, whenever you call `mix deps.update`, we will go the latest ref possible in a given branch/tag/rev. I am not sure if we can make it work for `mix deps.get` (which by definition will use whatever is in a lock).

josevalim replied:
Alternatively, we can do the new checks you have added only if you specify a tag or a revision but not if you specify a branch but I am not sure I like this.

josevalim commented on lib/mix/lib/mix/scm/git.ex:
```diff
@@ -123,7 +124,7 @@ defmodule Mix.SCM.Git do
   end
 
   defp do_checkout(opts) do
-    ref = get_lock_rev(opts[:lock]) || get_opts_rev(opts)
```
To recap what we have discussed, I think the issue is all here. We should not trust the lock if the lock options are out of date, so this should rather be something like:

    ref =  get_lock_rev_if_recent(...) || get_opts_rev(opts)

This should make the rev be updated in the scenario where the bug was reported.

whatyouhide commented:
@josevalim updated with a9eb435, let me know how it looks.

whatyouhide commented:
@josevalim ok, updated again, this time I think I got it right :). I introduced `lock_outdated?` which just looks if the repos are different or if the options from the lock are different than the ones in `mix.exs`, and used that in both the new `get_lock_rev_if_lock_up_to_date` function and in the `lock_status` function. There is some duplication because in both places we check if `opts[:lock]` is `{:git, ...}` or not, but it's reasonable I think.

A nitpick: `lock_outdated?` returns `true` even if `opts[:lock]` is something other than `nil` or `{:git, ...}`; for these cases, `lock_status` correctly returns `:mismatch`, but I'm not worried in this case because we're only calling `lock_outdated?` from "safe" places where we already know the value of `opts[:lock]` (I'm talking up in abstraction, in `scm.update`/`scm.checkout`, where we already called `lock_status`).

Wdyt?

josevalim commented on lib/mix/lib/mix/scm/git.ex:
```diff
@@ -229,4 +236,13 @@ defmodule Mix.SCM.Git do
+  defp lock_outdated?(opts) do
+    case opts[:lock] do
+      {:git, repo, _rev, lock_opts} ->
+        not git_repos_match?(repo, opts[:git]) or lock_opts != get_lock_opts(opts)
```
what if we call it `lock_up_to_date?` so we can write it as:

    git_repos_match?(repo, opts[:git]) and lock_opts == get_lock_opts(opts)

I think it will be clearer than the code today? Thoughts?

whatyouhide replied:
Agreed.

whatyouhide commented:
@josevalim refactored `lock_up_to_date?`.

josevalim commented:
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:
